src/commands/force/source/beta/pull.ts

```diff
@@ -0,0 +1,185 @@
+    wait: flags.minutes({
+      char: 'w',
+      default: Duration.minutes(33),
+      min: Duration.minutes(0), // wait=0 means deploy is asynchronous
+      description: messages.getMessage('flags.waitLong'),
+    }),
+  };
+
+  protected static requiresUsername = true;
+  protected static requiresProject = true;
+  protected hidden = true;
+  protected readonly lifecycleEventNames = ['preretrieve', 'postretrieve'];
+  protected tracking: SourceTracking;
+  protected retrieveResult: RetrieveResult;
+  protected deleteFileResponses: FileResponse[];
+
+  public async run(): Promise<PullResponse[]> {
+    await this.preChecks();
+    await this.retrieve();
+    // do not parallelize delete and retrieve...we only get to delete IF retrieve was successful
+    await this.doDeletes(); // deletes includes its tracking file operations
+    await this.updateTrackingFilesWithRetrieve();
+    this.ux.stopSpinner();
+
+    return this.formatResult();
+  }
+
+  protected async preChecks(): Promise<void> {
+    // checks the source tracking file version and throws if they're toolbelt's old version
+    throwIfInvalid({
+      org: this.org,
+      projectPath: this.project.getPath(),
+      toValidate: 'plugin-source',
+      command: replaceRenamedCommands('force:source:pull'),
+    });
+
+    this.ux.startSpinner('Loading source tracking information');
+    this.tracking = await SourceTracking.create({
+      org: this.org,
+      project: this.project,
+    });
+
+    await this.tracking.ensureRemoteTracking(true);
+    this.ux.setSpinnerStatus('Checking for conflicts');
+
+    if (!this.flags.forceoverwrite) {
+      processConflicts(await this.tracking.getConflicts(), this.ux, messages.getMessage('sourceConflictDetected'));
+    }
+  }
+
+  protected async doDeletes(): Promise<void> {
+    this.ux.setSpinnerStatus('Checking for deletes from the org and updating source tracking files');
+    const changesToDelete = await this.tracking.getChanges<SourceComponent>({
+      origin: 'remote',
+      state: 'delete',
+      format: 'SourceComponent',
+    });
+    this.deleteFileResponses = await this.tracking.deleteFilesAndUpdateTracking(changesToDelete);
+  }
+
+  protected async updateTrackingFilesWithRetrieve(): Promise<void> {
+    this.ux.setSpinnerStatus('Updating source tracking files');
+
+    const successes = this.retrieveResult
+      .getFileResponses()
+      .filter((fileResponse) => fileResponse.state !== ComponentStatus.Failed);
+
+    await Promise.all([
+      // commit the local file successes that the retrieve modified
+      this.tracking.updateLocalTracking({
+        files: successes.map((fileResponse) => fileResponse.filePath).filter(Boolean),
+      }),
+      this.tracking.updateRemoteTracking(
+        successes.map(({ state, fullName, type, filePath }) => ({ state, fullName, type, filePath })),
+        true // skip polling because it's a pull
+      ),
+    ]);
+  }
+
+  protected async retrieve(): Promise<void> {
```

Comment:
would it make sense to change this to `pull` instead? It's doing a little more than just a retrieve and might confuse the next person coming along. Same for push/deploy